egs/librispeech/asr/nnlm/main.py

```diff
@@ -0,0 +1,331 @@
+                    help='number of hidden units per layer')
+parser.add_argument('--nlayers', type=int, default=2, help='number of layers')
+parser.add_argument('--lr',
+                    type=float,
+                    default=20,
+                    help='initial learning rate')
+parser.add_argument('--clip',
+                    type=float,
+                    default=0.25,
+                    help='gradient clipping')
+parser.add_argument('--epochs', type=int, default=40, help='upper epoch limit')
+parser.add_argument('--batch_size',
+                    type=int,
+                    default=30,
+                    metavar='N',
+                    help='batch size')
+parser.add_argument('--bptt', type=int, default=35, help='sequence length')
+parser.add_argument('--dropout',
+                    type=float,
+                    default=0.2,
+                    help='dropout applied to layers (0 = no dropout)')
+parser.add_argument('--tied',
+                    action='store_true',
+                    help='tie the word embedding and softmax weights')
+parser.add_argument('--seed', type=int, default=1111, help='random seed')
+parser.add_argument('--cuda', action='store_true', help='use CUDA')
+parser.add_argument('--log-interval',
+                    type=int,
+                    default=200,
+                    metavar='N',
+                    help='report interval')
+parser.add_argument('--save',
+                    type=str,
+                    default='model.pt',
+                    help='path to save the final model')
+parser.add_argument('--onnx-export',
+                    type=str,
+                    default='',
+                    help='path to export the final model in onnx format')
+
+parser.add_argument(
+    '--nhead',
+    type=int,
+    default=2,
+    help='the number of heads in the encoder/decoder of the transformer model')
+parser.add_argument('--dry-run',
+                    action='store_true',
+                    help='verify the code and the model')
+
+args = parser.parse_args()
+
+# Set the random seed manually for reproducibility.
+torch.manual_seed(args.seed)
+if torch.cuda.is_available():
+    if not args.cuda:
+        print(
+            "WARNING: You have a CUDA device, so you should probably run with --cuda"
+        )
+
+device = torch.device("cuda" if args.cuda else "cpu")
+
+###############################################################################
+# Load data
+###############################################################################
+
+corpus = data.Corpus(args.data)
+
+# Starting from sequential data, batchify arranges the dataset into columns.
+# For instance, with the alphabet as the sequence and batch size 4, we'd get
+# ┌ a g m s ┐
+# │ b h n t │
+# │ c i o u │
+# │ d j p v │
+# │ e k q w │
+# └ f l r x ┘.
+# These columns are treated as independent by the model, which means that the
+# dependence of e. g. 'g' on 'f' can not be learned, but allows more efficient
```

Comment:
So is the data treated as one long sequence, rather than a bunch of independent sentences?
I would have thought for ASR applications, the independent-sentences approach might make  more sense.

Reply:
No, training text is not treated as a long sequence. I have modified the data preparation method so that each piece of text is treated independently.  Sorry to forget to delete these unrelated original comments. 
By the way, I am refactoring the training pipeline according these reviews. Temporarily, [a new dataset class is located here](https://github.com/glynpu/snowfall/blob/88e0d49d559860134bfdf244b38bf25c84fa2c56/egs/librispeech/asr/nnlm/local/dataset.py#L51), which handle training text one by one and then batchfy them independtly in [CollateFunc](https://github.com/glynpu/snowfall/blob/88e0d49d559860134bfdf244b38bf25c84fa2c56/egs/librispeech/asr/nnlm/local/dataset.py#L15).
```
        with open(text_file, 'r') as f:
            # a line represent a piece of text, e.g.
            # DELAWARE IS NOT AFRAID OF DOGS
            for line in f:
                text = line.strip().split()
                assert len(text) > 0
                text_id = self.text2id(text)
                # token_id format:
                # <bos_id> token_id token_id token_id *** <eos_id>
                token_id = self.text_id2token_id(text_id)
                self.data.append(token_id)
```